// convex/dist/esm/server/impl/storage_impl.js
"use strict";
import { performAsyncSyscall, performJsSyscall } from "./syscall.js";
import { validateArg } from "./validate.js";
export function setupStorageReader() {
  return {
    getUrl: async (storageId) => {
      validateArg(storageId, 1, "getUrl", "storageId");
      return await performAsyncSyscall("storageGetUrl", { storageId });
    },
    getMetadata: async (storageId) => {
      return await performAsyncSyscall("storageGetMetadata", { storageId });
    }
  };
}
export function setupStorageWriter() {
  const reader = setupStorageReader();
  return {
    generateUploadUrl: async () => {
      return await performAsyncSyscall("storageGenerateUploadUrl", {});
    },
    delete: async (storageId) => {
      await performAsyncSyscall("storageDelete", { storageId });
    },
    getUrl: reader.getUrl,
    getMetadata: reader.getMetadata
  };
}
export function setupStorageHttpWriter() {
  const writer = setupStorageWriter();
  return {
    ...writer,
    store: async (request) => {
      return await performJsSyscall("storage/storeFile", {
        request
      });
    },
    get: async (storageId) => {
      return await performJsSyscall("storage/getFile", {
        storageId
      });
    }
  };
}
